src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Common/IProjectContextInfoExtensions.cs

```diff
@@ -48,7 +52,47 @@ public static class IProjectContextInfoExtensions
 
             using (INuGetProjectManagerService projectManager = await GetProjectManagerAsync(serviceBroker, cancellationToken))
             {
-                return await projectManager.GetInstalledPackagesAsync(new string[] { projectContextInfo.ProjectId }, cancellationToken);
+                IReadOnlyDictionary<string, IReadOnlyCollection<IPackageReferenceContextInfo>> dictionary =
+                    await projectManager.GetInstalledPackagesAsync(new string[] { projectContextInfo.ProjectId }, cancellationToken);
+
+                if (dictionary.TryGetValue(projectContextInfo.ProjectId, out IReadOnlyCollection<IPackageReferenceContextInfo> packages))
+                {
+                    return packages;
+                }
+
+                return Array.Empty<IPackageReferenceContextInfo>();
+            }
+        }
+
+        /// <summary>
+        /// Gets installed packages for each project, and returns a dictionary mapping <see cref="IProjectContextInfo.ProjectId"/> to installed
+        /// <see cref="IPackageReferenceContextInfo"/>.
+        /// </summary>
+        /// <param name="projectContextInfos">Projects with a <see cref="IProjectContextInfo.ProjectId"/> to be searched.</param>
+        /// <param name="serviceBroker"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>Dictionary of <see cref="IProjectContextInfo.ProjectId"/> to installed <see cref="IPackageReferenceContextInfo"/>;
+        /// otherwise, an empty dictionary.</returns>
+        public static async ValueTask<IReadOnlyDictionary<string, IReadOnlyCollection<IPackageReferenceContextInfo>>> GetInstalledPackagesAsync(
+            this IReadOnlyCollection<IProjectContextInfo> projectContextInfos,
```

Comment:
I think this can be as follows, thereby simplifying places where you `.ToList()` before calling this.
```C#
this IEnumerable<IProjectContextInfo> projectContextInfos,
```